src/3rd_party_config.c

```diff
@@ -0,0 +1,296 @@
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "3rd_party_internal.h"
+#include "config.h"
+#include "lib/config_parser.h"
+#include "lib/list.h"
+#include "lib/sys.h"
+#include "swupd.h"
+#include <errno.h>
+#include <stdlib.h>
+
+#ifdef THIRDPARTY
+
+#define REPO_CONFIG_DIR "3rd_party"
+#define REPO_CONFIG_CONFIG_FILE "repo.ini"
+#define MAX_REPO_ALLOWED_FIELDS 2
+#define MAX_GLOBAL_ALLOWED_FIELDS 2
+#define MAX_FILE_PATH 1000
+
+/** @brief Storing global key,value pairs. */
+struct global {
+	void *value1;
+	void *value2;
+};
+
+/** @brief Internal Memory representation of a repo.  */
+struct repo {
+	char *repo_name;
+	char *url;
+};
+
+/** @brief List of repos. */
+struct list *repos_kv;
+
+/** @brief We populate this once every swupd init and use it many times. */
+static char *repo_config_file_path;
+
+/**
+ * @brief This function is called by the repo ini parser It is called
+ * once for every line which is successfully parsed with a key,value,section
+ * Once a section is completely read, it is stored in the global lists.
+ *
+ * @param section A string containing section of the file
+ * @param key A string containing key
+ * @param value A string containing value
+ *
+ * @returns True for every match of section with REPO or GLOBAL, false otherwise
+ */
+bool parse_key_values(char *section, char *key, char *value, void UNUSED_PARAM *data)
+{
+	static int repo_field_count = 0;
+	static int global_field_count = 0;
+	bool ret = false;
+
+	if (!strncmp(section, "REPO", strlen(section))) {
+		if (key && value) {
+			static struct repo *repo_var;
+			if (repo_field_count == 0) {
+				repo_var = (struct repo *)malloc(sizeof(struct repo));
+			}
+			if (!strncmp(key, "repo-name", strlen(key))) {
+				repo_var->repo_name = strdup_or_die(value);
+				repo_field_count++;
+				ret = true;
+			}
+
+			if (!strncmp(key, "repo-url", strlen(key))) {
+				repo_var->url = strdup_or_die(value);
+				repo_field_count++;
+				ret = true;
+			}
+
+			if (repo_field_count == MAX_REPO_ALLOWED_FIELDS) {
+				repos_kv = list_append_data(repos_kv, repo_var);
+				repo_field_count = 0;
+			}
+		}
+	} else if (!strncmp(section, "GLOBAL", strlen(section))) {
+		if (key && value) {
+			static struct global *global_var;
+			if (global_field_count > 2) {
+				return false;
+			}
+
+			if (!strncmp(key, "value1", strlen(key))) {
+				global_var->value1 = strdup_or_die(value);
+				global_field_count++;
+				ret = true;
+			}
+
+			if (!strncmp(key, "value2", strlen(key))) {
+				global_var->value2 = strdup_or_die(value);
+				global_field_count++;
+				ret = true;
+			}
+		}
+	}
+	return ret;
+}
```

Comment:
I think I have reservations with this schema as you dont really know what section you are really talking. What if we tomorrow we have to add another sub-section inside repo section. We might not be able to tell the difference between the two

Reply:
I don't see why you would need another subsection since you can have as many key/value pairs as you want. Besides INI files don't have subsections as far as I know. We are sticking to the standard INI format.

Reply:
So I think I misunderstood a little..how ini file is represented. Your understanding is consistent with Otavio's. Changing this and addressing your concerns in next fix.